### ISSUE TITLE
TypeError when trying to mesolve with single function callback

### ISSUE THREAD

**Issue by Phyks** [qutip/qutip#505]
Hi,

According to the `mesolve` doc, 
> For problems with time-dependent problems H and c_ops can be callback functions that takes two arguments, time and args, and returns the Hamiltonian or Liouvillian for the system at that point in time (callback format).

Then, I would have expected something very basic such as
```python
import numpy as np
import qutip

a = qutip.destroy(40)
tlist = np.arange(0, 100, 10)
results = qutip.mesolve(lambda t, args: a,
                        qutip.coherent(40, 2),
                        tlist,
                        [],
                        [],
                        progress_bar=True)
```
to be working.

Instead, it gives (full traceback)
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-116-e35e17a7e3ab> in <module>()
      6                         [],
      7                         [],
----> 8                         progress_bar=True)

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/qutip/mesolve.py in mesolve(H, rho0, tlist, c_ops, e_ops, args, options, progress_bar)
    319                             types.BuiltinFunctionType, partial)):
    320             res = _sesolve_func_td(H, rho0, tlist,
--> 321                                    e_ops, args, options, progress_bar)
    322         else:
    323             res = _sesolve_const(H, rho0, tlist,

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/qutip/sesolve.py in _sesolve_func_td(H_func, psi0, tlist, e_ops, args, opt, progress_bar)
    565     #
    566     return _generic_ode_solve(r, psi0, tlist, e_ops, opt, progress_bar, norm,
--> 567                               dims=psi0.dims)
    568 
    569 

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/qutip/sesolve.py in _generic_ode_solve(r, psi0, tlist, e_ops, opt, progress_bar, state_norm_func, dims)
    656 
    657         if t_idx < n_tsteps - 1:
--> 658             r.integrate(r.t + dt[t_idx])
    659 
    660     progress_bar.finished()

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/scipy/integrate/_ode.py in integrate(self, t, step, relax)
    406             self._y, self.t = mth(self.f, self.jac or (lambda: None),
    407                                 self._y, self.t, t,
--> 408                                 self.f_params, self.jac_params)
    409         except SystemError:
    410             # f2py issue with tuple returns, see ticket 1187.

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/scipy/integrate/_ode.py in run(self, f, jac, y0, t0, t1, f_params, jac_params)
    863         args = ((f, jac, y0, t0, t1) + tuple(self.call_args) +
    864                 (f_params, jac_params))
--> 865         y1, t, istate = self.runner(*args)
    866         if istate < 0:
    867             warnings.warn(self.__class__.__name__ + ': ' +

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/qutip/cy/spmatfuncs.pyx in qutip.cy.spmatfuncs.cy_ode_psi_func_td (qutip/cy/spmatfuncs.c:4445)()

/home/phyks/.local/share/virtualenvs/physique/lib/python3.5/site-packages/qutip/cy/spmatfuncs.pyx in qutip.cy.spmatfuncs.cy_ode_psi_func_td (qutip/cy/spmatfuncs.c:4289)()

TypeError: Cannot convert csr_matrix to numpy.ndarray
```

Is it expected? (in this case the doc may be misleading)

I tried to trace it down, but could not get any luck for the moment.

EDIT: Actually, it seems that in such a case, one has to pass a density matrix and not a state as initial state, and that the liouvillian callback should return a `.data` item.

**Comment by nonhermitian** [comment]
Will take a look next week.

**Comment by nonhermitian** [comment]
Fixed in pull #537.
